### PR TITLE
Test level retry

### PR DESCRIPTION
**Pull request by grouma**
Adds test level retry logic.

Partly addresses https://github.com/dart-lang/test/issues/441.

Group level and annotation support will follow.

### CODE REVIEW

File: lib/src/backend/metadata.dart [hunk 1]
```diff
@@ -193,6 +200,7 @@ class Metadata {
         _skip = skip,
         _verboseTrace = verboseTrace,
         _chainStackTraces = chainStackTraces,
+        _retry = retry,
```
Comment (nex3) [inline lib/src/backend/metadata.dart]: Add a [`RangeError.checkNotNegative()`](https://api.dartlang.org/stable/1.23.0/dart-core/RangeError/checkNotNegative.html) call to the body as well. Same for `new Metadata.parse()`.

File: lib/src/runner/engine.dart [hunk 2]
```diff
@@ -340,8 +340,11 @@ class Engine {
   ///
   /// If [countSuccess] is `true` (the default), the test is put into [passed]
   /// if it succeeds. Otherwise, it's removed from [liveTests] entirely.
+  ///
+  /// [attempt] is used to determine if this test should be retried, defaults to
+  /// 1.
```
Comment (nex3) [inline lib/src/runner/engine.dart]: Maybe "The [attempt] argument is only passed when [_runLiveTest] calls itself recursively. It indicates the number of times the test has been run, including the current time. It defaults to 1."

File: lib/src/runner/engine.dart [hunk 3]
```diff
@@ -340,8 +340,11 @@ class Engine {
...
   Future _runLiveTest(LiveSuiteController suiteController, LiveTest liveTest,
-      {bool countSuccess: true}) async {
+      {bool countSuccess: true, int attempt: 1}) async {
```
Comment (nex3) [inline lib/src/runner/engine.dart]: We generally avoid setting default values for non-boolean arguments, since it makes it difficult for callers to explicitly specify "the default value". Even in cases like this, where no callers are doing that, I like to use `??` in the body for consistency.

File: lib/src/runner/live_suite_controller.dart [hunk 4]
```diff
@@ -106,8 +106,12 @@ class LiveSuiteController {
   /// is `true` (the default), the test is put into [passed] if it succeeds.
   /// Otherwise, it's removed from [liveTests] entirely.
   ///
+  /// Likewise, if [countFailure] is `true` (the default), the test is put into
+  /// [failed] if it fails. Otherwise, it's removed from [liveTests] entirely.
```
Comment (nex3) [inline lib/src/runner/live_suite_controller.dart]: Another way to handle this would be to treat all retries as part of the same test, and just have that test not complete until all retries are complete—you'd probably handle that in the Invoker rather than the Engine.

If we decide to go with the `countFailures` approach, I think that means a breaking change to the JSON reporter protocol—it does have a notion of "hidden" tests currently, but the documentation guarantees that "only successful tests will be hidden" which won't be true after this. That indicates to me that a single-test approach would be less disruptive.

Comment (grouma) [inline lib/src/runner/live_suite_controller.dart]: I played around with trying to move the logic into the Invoker but was unfortunately unsuccessful. Ensuring the errors are printed, dealing with the timeout, and correctly setting the state resulted in a lot of additional logic. I quite like the idea of simply cloning the test and reusing all the existing logic. That being said I understand your concerns so lets chat in person and see if the difficulty of moving the logic into the Invoker was simply a result of my lack of experience with this code base.

**Comment by grouma** [comment]
I took a second crack at it and moved the logic over to the invoker. It was a lot more straightforward than my initial attempt. All tests pass; admittedly not many cover the retry branch and therefore I'm nervous I'm missing some edge conditions, especially around the timeout logic.

File: lib/src/backend/invoker.dart [hunk 5]
```diff
@@ -348,6 +352,12 @@ class Invoker {
         await _outstandingCallbacks.noOutstandingCallbacks;
         if (_timeoutTimer != null) _timeoutTimer.cancel();
 
+        if (liveTest.state.result != Result.success &&
+            _runCount < liveTest.test.metadata.retry + 1) {
+          _onRun();
```
Comment (natebosch) [inline lib/src/backend/invoker.dart]: have you explored what happens to the stack trace for these tests? I'm curious how things look when we go through the same Chain.capture() multiple tmes

**Comment by grouma** [comment]
Yes @nex3 called it out in her initial review. It is dependent on which approach we take. I have a pending change with that functionality.

File: lib/src/backend/invoker.dart (hunk 5 above)
Comment (nex3) [inline lib/src/backend/invoker.dart]: `Chain.capture()` is a no-op if it's called within a capture zone.

**Comment by nex3** [comment]
Does this work for `LoadSuite`s? If so, can you add a test?

**Comment by grouma** [comment]
This doesn't work with LoadSuite tests. It'll require some extra logic.

File: lib/src/backend/metadata.dart [hunk 6]
```diff
@@ -193,10 +200,12 @@ class Metadata {
...
         tags = new UnmodifiableSetView(tags == null ? new Set() : tags.toSet()),
         onPlatform =
             onPlatform == null ? const {} : new UnmodifiableMapView(onPlatform),
         forTag = forTag == null ? const {} : new UnmodifiableMapView(forTag) {
+    if (retry != null) RangeError.checkNotNegative(retry);
```
Comment (grouma) [inline lib/src/backend/metadata.dart]: Done.

**Comment by grouma** [comment]
@nex3 can you take a second look at this when you get a chance.

File: lib/src/backend/invoker.dart [hunk 7]
```diff
@@ -348,6 +352,14 @@ class Invoker {
         await _outstandingCallbacks.noOutstandingCallbacks;
         if (_timeoutTimer != null) _timeoutTimer.cancel();
 
+        if (liveTest.state.result != Result.success &&
+            _runCount < liveTest.test.metadata.retry + 1) {
+          _controller
+              .message(new Message.retry("Retry: ${liveTest.test.name}"));
```
Comment (nex3) [inline lib/src/backend/invoker.dart]: Are retry messages really worth a separate message type? We only have one for skips because we really want to print them in yellow text to match skipped tests.

File: lib/src/backend/metadata.dart (hunk 6 above)
Comment (nex3) [inline lib/src/backend/metadata.dart]: Looks like you may have missed an upload for this?

File: lib/src/runner/live_suite_controller.dart [hunk 8]
```diff
@@ -128,6 +128,7 @@ class LiveSuiteController {
         _failed.add(liveTest);
       } else if (countSuccess) {
         _passed.add(liveTest);
+        _failed.remove(liveTest);
```
Comment (nex3) [inline lib/src/runner/live_suite_controller.dart]: Add a comment here explaining why a failed test might start succeeding.

File: lib/src/runner/parse_metadata.dart [hunk 9]
```diff
@@ -60,6 +60,7 @@ class _Parser {
     var skip;
     Map<PlatformSelector, Metadata> onPlatform;
     Set<String> tags;
+    var retry;
```
Comment (nex3) [inline lib/src/runner/parse_metadata.dart]: Nit: `int` (`skip` is `var` because it can be a string or a bool)

File: test/runner/runner_test.dart [hunk 10]
```diff
@@ -292,6 +292,82 @@ $_usage""");
...
+            }, retry: 5);
+          }
+          """)
+          .create();
+
+      var test = runTest(["test.dart"]);
+      test.stdout.expect(consumeThrough(contains("+1: All tests passed!")));
+      test.shouldExit(0);
+    });
+  });
```
Comment (nex3) [inline test/runner/runner_test.dart]: These should probably be in `retry_test.dart`.

File: test/runner/runner_test.dart [hunk 11]
```diff
@@ -292,6 +292,82 @@ $_usage""");
...
+
+          import 'package:test/test.dart';
+
+          int attempt = 0;
+          void main() {
+            test("failure", () {
+            attempt++;
+            if (attempt != 2){
+             throw new TestFailure("oh no");
+            }
```
Comment (nex3) [inline test/runner/runner_test.dart]: A simpler way to do this might be to print a message each time the test runs, and use stream matchers to assert that the message is only printed once.

File: test/runner/runner_test.dart [hunk 12]
```diff
@@ -292,6 +292,82 @@ $_usage""");
...
+              "test.dart",
+              """
+          import 'dart:async';
+
+          import 'package:test/test.dart';
+
+          int attempt = 0;
+          void main() {
+            test("failure", () {
+             attempt++;
```
Comment (nex3) [inline test/runner/runner_test.dart]: Nit: +2 indentation within `test()` (also below).

File: test/runner/runner_test.dart [hunk 13]
```diff
@@ -292,6 +292,82 @@ $_usage""");
...
+          .file(
+              "test.dart",
+              """
+          import 'dart:async';
+
+          import 'package:test/test.dart';
+
+          int attempt = 0;
+          void main() {
+            test("failure", () {
```
Comment (nex3) [inline test/runner/runner_test.dart]: This should probably have a different name, since it succeeds :wink:.

File: test/runner/timeout_test.dart [hunk 14]
```diff
@@ -59,6 +59,36 @@ void main() {
...
+    runCount++;
+    if (runCount <=2) {
+      await new Future.delayed(new Duration(milliseconds: 1000));
+    }
+  }, retry: 3);
+}
+''')
+        .create();
+
+    var test = runTest(["--timeout=0.016x", "test.dart"]);
```
Comment (nex3) [inline test/runner/timeout_test.dart]: It's probably cleaner to set a time-based timeout here, rather than a factor-based one.

File: lib/src/backend/invoker.dart (hunk 7 above)
Comment (grouma) [inline lib/src/backend/invoker.dart]: My thought process is that we should print the retries in yellow as well. Most teams would likely want to reduce flakiness as much as possible, so having them stand out would be beneficial.

File: lib/src/backend/metadata.dart (hunk 6 above)
Comment (grouma) [inline lib/src/backend/metadata.dart]: Done here and elsewhere now :P

File: lib/src/runner/live_suite_controller.dart (hunk 8 above)
Comment (grouma) [inline lib/src/runner/live_suite_controller.dart]: Done.

File: lib/src/runner/parse_metadata.dart (hunk 9 above)
Comment (grouma) [inline lib/src/runner/parse_metadata.dart]: Done.

File: test/runner/runner_test.dart (hunk 12 above)
Comment (grouma) [inline test/runner/runner_test.dart]: Done. Wish there was a way to specify inline code blocks that were formatted better :(

File: test/runner/runner_test.dart (hunk 13 above)
Comment (grouma) [inline test/runner/runner_test.dart]: Done.

File: test/runner/runner_test.dart (hunk 11 above)
Comment (grouma) [inline test/runner/runner_test.dart]: Noted. Gonna leave this as is though.

File: test/runner/timeout_test.dart (hunk 14 above)
Comment (grouma) [inline test/runner/timeout_test.dart]: Done.

File: test/runner/runner_test.dart (hunk 10 above)
Comment (grouma) [inline test/runner/runner_test.dart]: I wasn't too sure on the distinction between runner test and the other various test file. I can see how runner_test can become a kitchen sink. I have moved these tests over.

File: lib/src/backend/invoker.dart (hunk 7 above)
Comment (nex3) [inline lib/src/backend/invoker.dart]: Unlike skips, there are valid reasons to mark a test as retried for the long term—for example, that's probably what we'll do for our browser tests, since browsers can sometimes just fail or time out arbitrarily. I don't want to force a super-visible message for those cases.

File: test/runner/runner_test.dart (hunk 12 above)
Comment (nex3) [inline test/runner/runner_test.dart]: Long ago there was a proposal for tagged template strings in Dart that could maybe work that way, but it never went anywhere :disappointed:. We could conceivably add a special case to the formatter to look for `"""// Dart` or something, but I doubt @munificent would go for that.

File: test/runner/runner_test.dart (hunk 10 above)
Comment (nex3) [inline test/runner/runner_test.dart]: Yeah, I've been trying to make more feature-specific files to avoid kitchen-sink-itis, but I haven't migrated all the old tests.

File: lib/src/backend/invoker.dart (hunk 7 above)
Comment (grouma) [inline lib/src/backend/invoker.dart]: Removed the special message type.

**Comment by grouma** [comment]
Discussed offline. I have added logic to ignore errors propagated from previous test runs and have added a  corresponding test.

File: lib/src/backend/invoker.dart [hunk 15]
```diff
@@ -357,7 +371,8 @@ class Invoker {
             // Use the invoker as a key so that multiple invokers can have different
             // outstanding callback counters at once.
             _counterKey: outstandingCallbacksForBody,
-            _closableKey: true
+            _closableKey: true,
+            #runCount: _runCount + 1
```
Comment (nex3) [inline lib/src/backend/invoker.dart]: I think it's a little confusing that this gets incremented both here and in `new Future()`. Can we increment it once before we called `runZonedWithValue()`?

File: test/runner/retry_test.dart [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+            "test.dart",
+            """
+          @Retry(3)
+
+          import 'dart:async';
+
+          import 'package:test/test.dart';
+
+
+          int attempt = 0;
```
Comment (nex3) [inline test/runner/retry_test.dart]: Nit: `var attempt`

We generally don't type variables that have initializers, since they'll be inferred anyway. Also below.

File: test/runner/retry_test.dart [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+              int attempt = 0;
+              Completer completer = new Completer();
+              void main() {
+                test("failure", () {
+                  attempt++;
+                  if (attempt == 1) {
+                    completer.future.then((_) => throw 'some error');
+                    throw new TestFailure("oh no");
+                  }
+                  completer.complete(null);
```
Comment (nex3) [inline test/runner/retry_test.dart]: I think this should only be completed during the second attempt. It would also be good to wait a little bit to ensure that the `throw` is reached.

File: test/runner/retry_test.dart [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+
+          import 'dart:async';
+
+          import 'package:test/test.dart';
+
+
+          int attempt = 0;
+          void main() {
+            test("failure", () {
+             attempt++;
```
Comment (nex3) [inline test/runner/retry_test.dart]: Nit: +1 indent (also below).

If it helps, you can copy these to a file, format that with `--indent 6`, and then copy them back here.

File: test/runner/retry_test.dart [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+                }, retry: 5);
+          }
+          """)
+          .create();
+
+      var test = runTest(["test.dart"]);
+      test.stdout.expect(consumeThrough(contains("+1: All tests passed!")));
+      test.shouldExit(0);
+    });
+  });
```
Comment (nex3) [inline test/runner/retry_test.dart]: We should also test what happens if a retry test fails after it reports success. Something like

```dart
void main() {
  var completer1 = new Completer();
  var completer2 = new Completer();
  test("first", () {
    completer1.future.then((_) {
      completer2.complete();
      throw "oh no";
    });
  });

  test("second", () async {
    completer1.complete();
    await completer2.future;
  });
}
```

File: lib/src/backend/invoker.dart (hunk 15 above)
Comment (grouma) [inline lib/src/backend/invoker.dart]: Done.

File: test/runner/retry_test.dart (hunk 18 above)
Comment (grouma) [inline test/runner/retry_test.dart]: Done.

File: test/runner/retry_test.dart (hunk 16 above)
Comment (grouma) [inline test/runner/retry_test.dart]: Done.

File: test/runner/retry_test.dart (hunk 17 above)
Comment (grouma) [inline test/runner/retry_test.dart]: This will only get completed in the second attempt since the throw will happen in the first attempt. I have added an await to ensure the throw is reached.

File: test/runner/retry_test.dart (hunk 19 above)
Comment (grouma) [inline test/runner/retry_test.dart]: Done.
